Generate chains for use with Viem/Wagmi

andreogle opened the pull request:
## Changes

1. Adds a new `viemConfig.chains()` function to generate chains for use in Viem/Wagmi. [`@wagmi/core/chains` is a proxy for `viem/chains`](https://wagmi.sh/core/chains#wagmicorechains). Addresses: https://github.com/api3dao/chains/issues/84

andreogle commented on src/viem-config.ts:
```diff
@@ -0,0 +1,41 @@
+  const exportedChains: Array<ViemChain> = [];
+
+  for (const name in viemChains) {
+    const viemChain = (viemChains as { [name: string]: ViemChain })[name];
```
@bbenligiray We'll need to decide what we want to do here as Viem doesn't support all of the same chains we have in our own `chains/` folder. We currently have 49 chains, with only 38 overlapping chains with Viem. i.e. this function will currently only output an array of length 38.

Currently, it re-uses whatever Viem has set for each chain (matching on chain ID) and then overwrites some URL fields to use what we've set.

I assume we'll want to add the remaining chains, but that will require adding more fields to whatever we currently have ([specifically adding fields like "nativeCurrency"](https://viem.sh/docs/clients/chains.html#custom-chains)). I wanted to get feedback first before going down that route, since it will be more work now (finding & setting the native currency/whatever else for 49 chains) and later (maintenance if these values ever change)

andreogle replied:
We could also not merge with the `viem` chains at all and just use `defineChain` for everything, but we could potentially lose some information. e.g [`contracts`, `formatters`, `serializers`](https://github.com/wagmi-dev/viem/blob/42d74b93c10c71ef8a8310f866788407f7472e0c/src/chains/definitions/celo.ts) etc. I'm not sure how big of a deal that is

andreogle replied:
I've been thinking about this a bit more and suspect we may we to use our own data, rather than merging with data from `viem/chains`

wkande replied:
@andreogle I have been using Viem in the Centurion app. It works _very_ well. Are any of our production  `@api3/chains` missing in **[Viem production chains list](https://github.com/wagmi-dev/viem/blob/main/src/chains/index.ts)** that you noticed?

wkande replied:
@andreogle I noticed this when I was in their docs. 

- https://viem.sh/docs/clients/chains.html#usage
- https://github.com/wagmi-dev/viem/blob/main/.github/CONTRIBUTING.md#chains

andreogle replied:
I did a quick check and these are the chains currently listed in our `chains/` directory, but not included in `viem/chains`

```
Excluded chain id:43288 name:Boba/Avalanche
Excluded chain id:56288 name:Boba/BNB
Excluded chain id:71401 name:Godwoken testnet
Excluded chain id:71402 name:Godwoken
Excluded chain id:2221 name:Kava testnet
Excluded chain id:200101 name:Milkomeda C1 testnet
Excluded chain id:2001 name:Milkomeda C1
Excluded chain id:31 name:RSK testnet
Excluded chain id:30 name:RSK
Excluded chain id:647 name:SX Network testnet
Excluded chain id:416 name:SX Network
```

I'm not sure how successful we'll be trying to get 11 new chains included in `viem/chains` 🤔 

> I have been using Viem in the Centurion app. It works _very_ well.

Glad to hear it 😎

bbenligiray replied:
> I've been thinking about this a bit more and suspect we may we to use our own data, rather than merging with data from viem/chains

I don't see any issues with this

andreogle commented:
PR update:

I have added `nativeCurrency` to each chain (replacing `symbol`). I copied the values from each chain defined by Viem here: https://github.com/wagmi-dev/viem/tree/main/src/chains/definitions. Where the chain didn't exist, I searched for values through the respective block explorer and through some extra Googling. I hope everything is correct.

`viemConfig.chains` no longer merges our definitions with the Viem definitions. Everything is now our definition.

I also renamed the following networks to more closely match the definitions from Viem:
1. Avalanche testnet -> Avalanche Fuji testnet
2. Gnosis Chain testnet -> Gnosis Chiado testnet

andreogle commented:
Also note, a lot of the chain `symbol` values have been changed. I'm not sure if these were being used or not, but the old values didn't seem correct to me

bbenligiray commented on chains/arbitrum-goerli-testnet.json:
```diff
@@ -2,7 +2,11 @@
   "id": "421613",
-  "symbol": "testETH",
+  "nativeCurrency": {
+    "name": "Arbitrum Goerli Ether",
```
I find this field a bit silly, can we just call it `X` when the symbol is `X` and `Testnet X` when the symbol is `testX`?

andreogle replied:
We could actually drop this field entirely and go back to `symbol` at the top level if we want to follow that pattern. When generating the viem compatible chain (`defineChain`) and for chains where `testnet: true`, it strips off the `test` prefix from the symbol and adds `"Testnet ". i.e. `testETH` becomes `Testnet ETH` automatically. I'll admit I'm not the biggest fan of the `nativeCurrency` object.

The only thing we need which doesn't currently exist is `decimals`

I'm also thinking of adding validation to ensure that for all chains where `testnet: true`, the symbol is prefixed with `test`. What do you think?

andreogle replied:
I guess we just need to decide if the pattern of having symbols prefixed with `test` is a pattern we want to enforce for all chains

bbenligiray replied:
> We could actually drop this field entirely and go back to symbol at the top level if we want to follow that pattern

Sounds good

> The only thing we need which doesn't currently exist is decimals

I doubt that we will ever have a chain with non-18 decimals

> I'm also thinking of adding validation to ensure that for all chains where testnet: true, the symbol is prefixed with test. What do you think?

Sounds good

> I guess we just need to decide if the pattern of having symbols prefixed with test is a pattern we want to enforce for all chains

I think that's a safe assumption

andreogle replied:
> I doubt that we will ever have a chain with non-18 decimals

Good point. For some reason, I thought there were a couple with < 18 decimals. I considered removing `decimals` and hardcoding it to 18 now, but this might lead to unintended bugs later on. I think it's safer to leave it in

bbenligiray replied:
I was suggesting that we hardcode it to 18 but they're equivalent solutions anyway

bbenligiray reviewed (commented):
Overall, I'm for being very opinionated and disregarding the official names/conventions in favor of optimizing the UX of @api3/chains users

andreogle commented on src/viem-config.ts:
```diff
@@ -0,0 +1,29 @@
+
+export function chains(): Chain[] {
+  return CHAINS.map((chain) => {
+    const viemChain = defineChain({
```
Just a personal style preference I normally use. I'll just return directly here because it's not adding much

andreogle commented on chains/arbitrum-goerli-testnet.json:
```diff
@@ -2,7 +2,11 @@
   "id": "421613",
-  "symbol": "testETH",
+  "nativeCurrency": {
+    "name": "Arbitrum Goerli Ether",
+    "symbol": "ETH",
```
Sure - I don't mind reverting the symbol chnages

andreogle commented on chains/avalanche-fuji-testnet.json:
```diff
@@ -1,8 +1,12 @@
 {
-  "name": "Avalanche testnet",
-  "alias": "avalanche-testnet",
+  "name": "Avalanche Fuji testnet",
```
![facepalm-really](https://github.com/api3dao/chains/assets/4114009/648a6d09-e921-4f33-ba57-cfcad3ac5034)

andreogle commented on src/viem-config.ts:
```diff
@@ -0,0 +1,29 @@
+      network: chain.alias,
+      nativeCurrency: chain.nativeCurrency,
+      rpcUrls: {
+        default: {
+          http: [chain.providerUrl],
+        },
+        public: {
+          http: [chain.providerUrl],
+        },
+      },
+      blockExplorers: {
+        default: {
+          name: 'Explorer',
+          url: chain.explorer.browserUrl,
+        },
+      },
```
I'm not quite sure I understand the difference between a "default" provider and a "public" provider. Also related to this: https://github.com/api3dao/chains/pull/108#discussion_r1368379592

bbenligiray replied:
To my understanding `public` is the URL of the official RPC that "the team" provides, and `default` is what we recommend the user to use. `default` must exist, while `public` may not exist (e.g. Ethereum). In a lot of the cases where `public` exists, it will also be `default`, except we sometimes recommend a third-party RPC (e.g. https://github.com/api3dao/chains/blob/ee80000efc629d4f646fe67c6bf00e2a83bf707b/chains/polygon.json#L7)

bbenligiray replied:
So in this case, I would only do
```
      rpcUrls: {
        default: {
          http: [chain.providerUrl],
        },
      },
```

andreogle replied:
Turns out viem requires both `default` and `public` as part of the `rpcUrls` object

bbenligiray reviewed (approved):
Agreed to remove `nativeCurrency`
